kv_connectors/llmd_fs_backend/src/csrc/storage/tensor_copy.cu

```diff
@@ -0,0 +1,547 @@
+
+    // Fetch or allocate thread-local staging buffer
+    StagingBufferInfo buf = get_thread_local_staging_buffer(total_bytes);
+    TORCH_CHECK(buf.size >= total_bytes,
+                "staging buffer too small: need ",
+                total_bytes,
+                " got ",
+                buf.size);
+
+    auto dtype = ref.dtype();
+    const int64_t total_elements =
+        static_cast<int64_t>(total_bytes / layout.elem_size);
+
+    // Wrap staging buffer as tensor view (no copy)
+    torch::Tensor result_cpu = torch::from_blob(buf.ptr,
+                                                {total_elements},
+                                                torch::TensorOptions()
+                                                    .dtype(dtype)
+                                                    .device(torch::kCPU)
+                                                    .pinned_memory(true));
+
+    auto* cpu_base = static_cast<uint8_t*>(buf.ptr);
+
+    // Check environment variable to determine copy method
+    bool use_kernel = get_env_flag("USE_KERNEL_COPY_WRITE", false);
+    bool is_put = true;
+
+    // Execute the copy operation
+    transfer_kv_blocks(cpu_base,
+                       src_tensors,
+                       block_ids_list,
+                       layout,
+                       stream,
+                       is_put,
+                       use_kernel);
+
+    // Reinterpret bfloat16 tensor as uint16_t for safe raw byte access (I/O or
+    // memcpy)
+    if (result_cpu.dtype() == torch::kBFloat16) {
+        result_cpu = result_cpu.view(torch::kUInt16);
+    }
+
+    return result_cpu;
+}
+
+//----------------------------------------------------------------------
+// Storage → GPU (GET)
+//----------------------------------------------------------------------
+
+bool copy_buffer_to_gpu_tensors(torch::Tensor cpu_buf,
+                                const std::vector<int64_t>& block_ids_list,
+                                const std::vector<torch::Tensor>& dst_tensors,
+                                int num_blocks_in_file,
+                                const c10::cuda::CUDAStream& stream) {
+    TORCH_CHECK(!dst_tensors.empty(), "Destination tensors list is empty");
+    const auto& ref = dst_tensors[0];
+    TORCH_CHECK(ref.is_contiguous(), "dst_tensors must be contiguous");
+    TORCH_CHECK(cpu_buf.is_contiguous(), "cpu buffer must be contiguous");
+
+    // Verify cpu_buf is pinned memory
+    TORCH_CHECK(cpu_buf.is_pinned(),
+                "cpu_buf must be pinned memory for kernel-based copy");
+
+    // Extract tensor geometry
+    auto layout =
+        CacheLayout::from_tensor(ref,
+                                 g_connector_config.block_axis,
+                                 g_connector_config.kv_before_blocks,
+                                 g_connector_config.layers_before_blocks);
+    const int num_layers = static_cast<int>(dst_tensors.size());
```

Comment:
you are only serving a single model, so this calculation is redundant.
Layout should be set once in a constructor of a wrapping class.